feat: support with clause

aceforeverd opened the pull request:
1. WITH clause
    Syntax and CTE rules follows Bigquery: https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#with_clause
2. fix doc of `*_cate_where`, close #2882

aceforeverd committed:
save changes

aceforeverd committed:
refactor

aceforeverd committed:
feat: with clause in batch mode

aceforeverd committed:
refactor

aceforeverd committed:
reactor: with clause

- lazy transforming with clause entries
- supports request mode (basicly)

aceforeverd committed:
feat(planner): support with clause

aceforeverd committed:
feat: optimize non-sense project(last join)

aceforeverd committed:
disalbe test

aceforeverd committed:
save changes

aceforeverd committed:
rm simplify optimize

aceforeverd committed:
fix tests

aceforeverd committed:
feat: with clause online

aceforeverd committed:
fix: with clause online

- fix: `count_cate` & `count_cate_where` support more key types

aceforeverd committed:
fix: Planner::ValidPlan

zhanghaohit commented on hybridse/include/base/fe_object.h:
```diff
@@ -25,6 +28,19 @@ class FeBaseObject {
+    void SetNodeId(size_t id) { node_id_ = id; }
+
+    std::vector<T*> data_;
+    size_t node_id_ = 0;
```
make these two members `private`?

aceforeverd replied:
`node_id_` is fine. `data_` is referenced elsewhere so should public

zhanghaohit commented on hybridse/include/node/sql_node.h:
```diff
@@ -579,10 +580,28 @@ class QueryNode : public SqlNode {
+
+class WithClauseEntry : public SqlNode {
+ public:
+    WithClauseEntry(std::string alias, QueryNode *query)
```
`const std::string& alias`

zhanghaohit commented on hybridse/src/plan/planner.cc:
```diff
@@ -547,6 +568,27 @@ int Planner::GetPlanTreeLimitCount(node::PlanNode *node) {
     return limit_cnt;
 }
 
+base::Status Planner::ValidPlan(node::PlanNode *node, bool is_primary_path) {
```
maybe name as `ValidRequestPlan`? this is for `request` mode only?

aceforeverd replied:
I'd take `ValidPlanForRequestMode`, plan node is no difference between batch and request mode

zhanghaohit commented on hybridse/src/udf/default_defs/count_by_category_def.cc:
```diff
@@ -260,7 +260,7 @@ void DefaultUdfLibrary::InitCountByCateUdafs() {
 
             @code{.sql}
-                SELECT count_cate_where(catagory, value, condition) OVER w;
+                SELECT count_cate_where(value, condition, category) OVER w;
```
u may also want to re-order the @param  doc parts

zhanghaohit replied:
all the `xx_cate_where` seem have the same issues

aceforeverd replied:
track in #2882, I won't fix all it here, except `count_cate_where` since it needed of some of my tests

aceforeverd committed:
refactor: node manager

aceforeverd committed:
fix: test in src & refactor Explain

Now it is preferred to get dependent tables from Physical Plan instead
of logic plan previously. The behavior has changed inside Explain due to
this, since a TablePlan logic plan node may not refer to a actual table,
but also refer to a CTE in a query contains WITH clause.

aceforeverd committed:
fix  & refactor: compile_engine_test

aceforeverd committed:
chore: format

aceforeverd committed:
refactor: GetDependentTables

aceforeverd committed:
feat!: deprecate GetDependentTables from PlanNode

aceforeverd committed:
sql_sdk_test: deployment test from yaml

aceforeverd committed:
fix: sql_sdk_test

aceforeverd committed:
test(yaml): load table data from csv

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into feat-with-clause

aceforeverd committed:
fix(sql_sdk_test): rm absl::BitGen

BitGen seems to core dump, reason unknown

aceforeverd committed:
sql_sdk_run

aceforeverd committed:
sql sdk run

aceforeverd committed:
Merge remote-tracking branch 'upstream/main' into feat-with-clause

aceforeverd committed:
fix conflicts

aceforeverd committed:
build: fix link on gcc < 9.1

aceforeverd committed:
build: fix link stdc++fs

aceforeverd committed:
feat: merge changes from aceforeverd:feat-icbc

aceforeverd committed:
test: fix index creation for deploy with SQL

aceforeverd committed:
chore: fix doc for *_cate_where

fixes #2882
